File: web-react/src/utils/limitEnforcement.test.ts
import { describe, expect, it } from 'vitest'
import zh from '@/locales/zh-CN/admin.json'
import en from '@/locales/en-US/admin.json'
import type { LimitEnforcement, LimitPanel } from '@/api/limitEnforcement'
import type { CapVerdict } from './limitEnforcement'
import { assessDeviceLimit, assessIPLimit, ipEnforcing, isMisleading } from './limitEnforcement'

function panel(over: Partial<LimitPanel> = {}): LimitPanel {
  return {
    panel_id: 1, name: 'de-fra', kind: 'xui', clients: 1,
    can_store_ip: true, can_store_device: true,
    ip_limit_enforcement: 'enforced',
    ...over,
  }
}

function d(over: Partial<LimitEnforcement> = {}): LimitEnforcement {
  return {
    ip_limit: 2, device_limit: 0, device_limit_enforced_anywhere: false,
    panels: [panel()],
    ...over,
  }
}

describe('ipEnforcing', () => {
  // Mirrors domain.IPLimitEnforcement.Enforcing(). If that Go method gains a
  // state, this must too — the two are one rule expressed twice.
  it('counts only the states where something actually happens', () => {
    expect(ipEnforcing('enforced')).toBe(true)
    // Windows: the client is disconnected but no IP ban follows. Still an
    // enforcement, just a weaker one.
    expect(ipEnforcing('disconnect_only')).toBe(true)
  })
  it('never counts an unread precondition as a working one', () => {
    for (const s of ['unknown', 'unsupported', 'disabled', 'not_installed', '']) {
      expect(ipEnforcing(s)).toBe(false)
    }
  })
})

describe('assessIPLimit', () => {
  it('says nothing when no cap is set', () => {
    expect(assessIPLimit(d({ ip_limit: 0 })).verdict).toBe('unlimited')
  })

  it('is exact when one panel with one client enforces', () => {
    const a = assessIPLimit(d())
    expect(a.verdict).toBe('exact')
    expect(a.enforcingRows).toBe(1)
  })

  // The cap is budgeted per (panel, email), so a user on three enforcing
  // panels can hold three times what the admin typed.
  it('multiplies by the number of enforcing panel-side rows', () => {
    const a = assessIPLimit(d({ panels: [
      panel({ panel_id: 1 }),
      panel({ panel_id: 2 }),
      panel({ panel_id: 3 }),
    ] }))
    expect(a.verdict).toBe('multiplied')
    expect(a.enforcingRows).toBe(3)
  })

  it('counts a second credential class on ONE panel as a second enforcer', () => {
    const a = assessIPLimit(d({ panels: [panel({ clients: 2 })] }))
    expect(a.verdict).toBe('multiplied')
    expect(a.enforcingRows).toBe(2)
  })

  // The rule that matters most: one hole makes the cap void, not merely
  // higher. The user routes everything through the panel that does not
  // enforce, so a "x2" reading would be actively wrong.
  it('is void, not multiplied, when any panel cannot store the field', () => {
    const a = assessIPLimit(d({ panels: [
      panel({ panel_id: 1 }),
      panel({ panel_id: 2, kind: 'sui', can_store_ip: false }),
    ] }))
    expect(a.verdict).toBe('void')
    expect(a.voidPanels.map(p => p.panel_id)).toEqual([2])
  })

  it('is void when a panel stores the value but fail2ban is off', () => {
    // The worst shape: the panel accepts the write and shows the number back,
    // so it looks configured while banning nobody.
    for (const state of ['disabled', 'not_installed']) {
      const a = assessIPLimit(d({ panels: [panel({ ip_limit_enforcement: state })] }))
      expect(a.verdict).toBe('void')
    }
  })

  it('separates an unread panel from a known-bad one', () => {
    for (const state of ['unknown', 'unsupported']) {
      const a = assessIPLimit(d({ panels: [panel({ ip_limit_enforcement: state })] }))
      expect(a.verdict).toBe('unknown')
      expect(a.voidPanels).toHaveLength(0)
      expect(a.unknownPanels).toHaveLength(1)
    }
  })

  it('reports an unreachable or deleted panel as unknown, never as enforcing', () => {
    for (const over of [{ panel_unreachable: true }, { panel_missing: true }]) {
      const a = assessIPLimit(d({ panels: [panel(over)] }))
      expect(a.verdict).toBe('unknown')
      expect(a.enforcingRows).toBe(0)
    }
  })

  // A hole outranks an unknown: the fleet-wide cap is already gone, and
  // reporting "we could not read one panel" would bury the certain finding
  // under the uncertain one.
  it('reports void ahead of unknown when both are present', () => {
    const a = assessIPLimit(d({ panels: [
      panel({ panel_id: 1, can_store_ip: false }),
      panel({ panel_id: 2, ip_limit_enforcement: 'unknown' }),
    ] }))
    expect(a.verdict).toBe('void')
  })

  it('says the user has no clients rather than claiming a working cap', () => {
    expect(assessIPLimit(d({ panels: [] })).verdict).toBe('no_clients')
  })
})

describe('assessDeviceLimit', () => {
  it('says nothing when no cap is set', () => {
    expect(assessDeviceLimit(d({ device_limit: 0 })).verdict).toBe('unlimited')
  })

  // The headline fact. Upstream counts a device only when a client app fetches
  // its subscription FROM THE PANEL, and PSP serves the subscriptions — so the
  // gate never fires for a PSP user on any panel version. This is not "no
  // panel happens to support it": no panel can.
  it('is never enforced, whatever the panels report', () => {
    const a = assessDeviceLimit(d({
      device_limit: 3,
      device_limit_enforced_anywhere: false,
      panels: [panel({ can_store_device: true })],
    }))
    expect(a.verdict).toBe('never_enforced')
  })

  it('falls back to the per-panel rules if a panel ever does enforce it', () => {
    // Guards the day the server flips the flag: the branch below must already
    // behave, rather than being written under pressure at that point.
    const a = assessDeviceLimit(d({
      device_limit: 3,
      device_limit_enforced_anywhere: true,
      panels: [panel({ panel_id: 1 }), panel({ panel_id: 2, can_store_device: false })],
    }))
    expect(a.verdict).toBe('void')
    expect(a.voidPanels.map(p => p.panel_id)).toEqual([2])
  })
})

describe('isMisleading', () => {
  it('flags exactly the verdicts where the typed number is not the enforced one', () => {
    const misleading: CapVerdict[] = ['multiplied', 'void', 'unknown', 'never_enforced']
    const honest: CapVerdict[] = ['unlimited', 'exact', 'no_clients']
    expect(misleading.every(isMisleading)).toBe(true)
    expect(honest.some(isMisleading)).toBe(false)
  })
})

// Every verdict the assessors can emit must have wording in BOTH bundles, and
// in BOTH halves: the summary shown under the field and the detail behind the
// click. The keys are computed at runtime, so a hint added with only one half
// renders as a raw key rather than failing to compile — and this repo has
// already shipped a missing en-US key that silently fell back to Chinese.
describe('locale coverage', () => {
  // Verdicts that render nothing, so they need no wording.
  const SILENT: CapVerdict[] = ['unlimited', 'exact']

  const KEYS: Record<Exclude<CapVerdict, 'unlimited' | 'exact'>, string> = {
    never_enforced: 'device_limit_inert',
    void: 'cap_void',
    unknown: 'cap_unknown',
    multiplied: 'cap_multiplied',
    no_clients: 'cap_no_clients',
  }

  // The fleet-wide hints on the create form share the same component and the
  // same key-stem convention, so they are held to the same rule.
  const CREATE_FORM_KEYS = ['limit_unsupported', 'ip_limit_unenforced']

  it('covers every verdict that renders wording', () => {
    expect(Object.keys(KEYS).length + SILENT.length).toBe(7)
  })

  it('has BOTH the summary and the detail, in both bundles', () => {
    for (const key of [...Object.values(KEYS), ...CREATE_FORM_KEYS]) {
      for (const [lang, bundle] of [['zh-CN', zh], ['en-US', en]] as const) {
        expect(bundle.users.field, `${lang} missing detail ${key}`).toHaveProperty(key)
        expect(bundle.users.field, `${lang} missing summary ${key}_short`).toHaveProperty(`${key}_short`)
      }
    }
  })

  // A summary that runs to a sentence defeats the point — it reflows the form
  // it was shortened to protect. Generous bound: this catches "somebody pasted
  // the detail in", not a word of drift.
  it('keeps summaries short enough to sit under a field', () => {
    for (const key of [...Object.values(KEYS), ...CREATE_FORM_KEYS]) {
      for (const [lang, bundle] of [['zh-CN', zh], ['en-US', en]] as const) {
        const text = (bundle.users.field as Record<string, string>)[`${key}_short`]
        expect(text.length, `${lang} ${key}_short is too long to be a summary: ${text}`)
          .toBeLessThanOrEqual(32)
      }
    }
  })

  it('keeps the placeholders each half interpolates', () => {
    for (const bundle of [zh, en]) {
      const f = bundle.users.field as Record<string, string>
      expect(f.cap_void).toContain('{{panels}}')
      expect(f.cap_unknown).toContain('{{panels}}')
      for (const ph of ['{{total}}', '{{rows}}', '{{typed}}']) {
        expect(f.cap_multiplied).toContain(ph)
      }
      // The summary carries only the number — the panel names are what the
      // detail exists for.
      expect(f.cap_multiplied_short).toContain('{{total}}')
    }
  })
})

File: internal/service/sharedclient/limit_enforcement_test.go
package sharedclient

import (
	"context"
	"errors"
	"testing"

	"github.com/KazuhaHub/passwall-sub-panel/internal/domain"
	"github.com/KazuhaHub/passwall-sub-panel/internal/ports"
)

// capPool hands a different adapter per panel, so one user can straddle a
// panel that carries the caps and one that does not — the situation the edit
// form has to describe and could not.
type capPool struct {
	ports.XUIPool
	byPanel map[int64]ports.XUIClient
}

func (p capPool) Get(id int64) (ports.XUIClient, error) {
	c, ok := p.byPanel[id]
	if !ok {
		return nil, errors.New("panel not registered")
	}
	return c, nil
}

// capClient declares exactly the capability set it is given.
type capClient struct {
	ports.XUIClient
	caps []ports.PanelCapability
}

func (c capClient) Capabilities() []ports.PanelCapability { return c.caps }

func bothCaps() capClient {
	return capClient{caps: []ports.PanelCapability{
		ports.CapabilityClientIPLimit, ports.CapabilityClientDeviceLimit,
	}}
}

// The connection caps are enforced per (panel, client email) while an admin
// types ONE number for the user, so the form's number and the enforced number
// are different quantities. This reports the facts that difference is made of.
func TestLimitEnforcement(t *testing.T) {
	clients := &fakeClients{byUser: []*domain.PSPClient{
		{ID: 1, UserID: 7, PanelID: 10},
		// Second credential class on the SAME panel: two panel-side rows, so
		// this panel alone enforces the typed cap twice.
		{ID: 2, UserID: 7, PanelID: 10, CredClass: 1},
		{ID: 3, UserID: 7, PanelID: 20},
		{ID: 4, UserID: 9, PanelID: 30}, // another user; must not leak in
	}}
	svc := New(clients, capPool{byPanel: map[int64]ports.XUIClient{
		10: bothCaps(),
		// S-UI's shape: carries neither cap. A user on this panel has no
		// enforced cap there at all.
		20: capClient{caps: []ports.PanelCapability{ports.CapabilityClientWrite}},
	}}, fakeNodes{})

	got, err := svc.LimitEnforcement(context.Background(), 7)
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 2 {
		t.Fatalf("panels = %d, want 2 (one row per PANEL, not per client)", len(got))
	}

	if got[0].PanelID != 10 || got[0].Clients != 2 {
		t.Fatalf("panel 10 = %+v, want 2 clients", got[0])
	}
	if !got[0].CanStoreIP || !got[0].CanStoreDevice {
		t.Fatalf("panel 10 should carry both caps: %+v", got[0])
	}
	if got[1].PanelID != 20 || got[1].Clients != 1 {
		t.Fatalf("panel 20 = %+v, want 1 client", got[1])
	}
	if got[1].CanStoreIP || got[1].CanStoreDevice {
		t.Fatalf("panel 20 carries neither cap, so both must be false: %+v", got[1])
	}
}

// A panel the pool will not hand over is UNKNOWN, not "cannot store". The two
// read differently to an operator and only one of them is actionable, and an
// unread precondition must never render as a working cap — the same rule the
// fail2ban verdict is held to.
func TestLimitEnforcementSeparatesUnreachableFromUncapable(t *testing.T) {
	clients := &fakeClients{byUser: []*domain.PSPClient{{ID: 1, UserID: 7, PanelID: 99}}}
	svc := New(clients, capPool{byPanel: map[int64]ports.XUIClient{}}, fakeNodes{})

	got, err := svc.LimitEnforcement(context.Background(), 7)
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 1 {
		t.Fatalf("panels = %d, want 1 — a panel we cannot read must be REPORTED, not dropped", len(got))
	}
	if !got[0].PanelUnreachable {
		t.Fatalf("want PanelUnreachable: %+v", got[0])
	}
	if got[0].CanStoreIP || got[0].CanStoreDevice {
		t.Fatalf("an unreachable panel must not claim a capability: %+v", got[0])
	}
}

// A user with no clients has nothing enforcing anything, and that is an empty
// list rather than an error: it is the ordinary state of a freshly created
// user, and an error here would put a red banner on the edit form.
func TestLimitEnforcementNoClients(t *testing.T) {
	svc := New(&fakeClients{}, capPool{}, fakeNodes{})
	got, err := svc.LimitEnforcement(context.Background(), 7)
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Fatalf("panels = %d, want 0", len(got))
	}
}
